lib/puppet/transaction.rb

```diff
@@ -287,15 +291,25 @@ def failed_dependencies?(resource)
     # explosion of edges, we also ended up reporting failures for containers
     # like class and stage.  This is undesirable; while just skipping the
     # output isn't perfect, it is RC-safe. --daniel 2011-06-07
-    suppress_report = (resource.class == Puppet::Type.type(:whit))
-
+    is_puppet_class = resource.class == Puppet::Type.type(:whit)
+    # With merge_dependency_warnings, notify about class dependency failures ... just once per class. TJK 2019-09-09
     s = resource_status(resource)
     if s && s.dependency_failed?
-      # See above. --daniel 2011-06-06
-      unless suppress_report then
-        s.failed_dependencies.find_all { |d| !(@failed_dependencies_already_notified.include?(d.ref)) }.each do |dep|
-          resource.notice _("Dependency %{dep} has failures: %{status}") % { dep: dep, status: resource_status(dep).failed }
-          @failed_dependencies_already_notified.add(dep.ref)
+      if @merge_dependency_warnings && is_puppet_class
+       	# Notes: Puppet::Resource::Status.failed_dependencies() is an Array of Puppet::Resource(s) and
+        #        Puppet::Resource.ref() calls Puppet::Resource.to_s() which is: "#{type}[#{title}]" and
+       	#        Puppet::Resource.resource_status(resource) calls Puppet::Resource.to_s()
+        class_dependencies_to_be_notified = (s.failed_dependencies.map(&:ref) - @failed_class_dependencies_already_notified.to_a)
+        class_dependencies_to_be_notified.each do |dep_ref|
+          resource.notice _("Class dependency %{dep} has failures: %{status}") % { dep: dep_ref, status: resource_status(dep_ref).failed }
+        end
+        @failed_class_dependencies_already_notified.merge(class_dependencies_to_be_notified)
+      else
+        if !@merge_dependency_warnings && !is_puppet_class
```

Comment:
could be replaced for better readability with: `unless @merge_dependency_warnings || is_puppet_class`